### ISSUE TITLE
equality of Gröbner bases

### ISSUE THREAD

**Issue by wdecker** [oscar-system/Oscar.jl#2019]
**Describe the bug**
I am not yet sure whether this is a bug, but in the example below the
two computed Gröbner bases are not considered equal despite the
fact that there elements are pairwise equal:

**To Reproduce**
Steps to reproduce the behavior, please provide a code snippet that triggers
the bug.
```
using Oscar
# Julia code
R, (x, y, z) = polynomial_ring(GF(32003), ["x", "y", "z"]);
f1 = x^2*y+169*y^21+151*x*y*z^10;
f2 = 6*x^2*y^4+x*z^14+3*z^24;
f3 = 11*x^3+5*x*y^10*z^10+2*y^20*z^10+y^10*z^20;
I = ideal(R, [f1, f2,f3]);
W = [10, 1, 1];
GB1 = groebner_basis_hilbert_driven(I, destination_ordering = lex(R), weights = W);
S, t = polynomial_ring(ZZ, "t")
hn = -t^75 + t^54 + t^51 + t^45 - t^30 - t^24 - t^21 + 1
GB2 = groebner_basis_hilbert_driven(I, destination_ordering = lex(R), weights = W, hilbert_numerator = hn);
GB1 == GB2
for i in 1:length(GB1)
 println(GB1.gens.S[i] == GB2.gens.S[i])
end
```

**Comment by wdecker** [comment]
@jankoboehm   Any idea?

**Comment by fieker** [comment]
On Wed, Mar 08, 2023 at 09:55:01AM -0800, Wolfram Decker wrote:
> **Describe the bug**
> I am not yet sure whether this is a bug, but in the example below the
> two computed Gröbner bases are not considered equal despite the
> fact that there elements are pairwise equal:
> 
> **To Reproduce**
> Steps to reproduce the behavior, please provide a code snippet that triggers
> the bug.
> ```
> using Oscar
> # Julia code
> R, (x, y, z) = polynomial_ring(GF(32003), ["x", "y", "z"]);
> f1 = x^2*y+169*y^21+151*x*y*z^10;
> f2 = 6*x^2*y^4+x*z^14+3*z^24;
> f3 = 11*x^3+5*x*y^10*z^10+2*y^20*z^10+y^10*z^20;
> I = ideal(R, [f1, f2,f3]);
> W = [10, 1, 1];
> GB1 = groebner_basis_hilbert_driven(I, destination_ordering = lex(R), weights = W);
> S, t = polynomial_ring(ZZ, "t")
> hn = -t^75 + t^54 + t^51 + t^45 - t^30 - t^24 - t^21 + 1
> GB2 = groebner_basis_hilbert_driven(I, destination_ordering = lex(R), weights = W, hilbert_numerator = hn);
> GB1 == GB2
> for i in 1:length(GB1)
>  println(GB1.gens.S[i] == GB2.gens.S[i])
> end
> ```
> 
There is no "==" for Groebner basis - it defaults to "===". My guess is
that normally, since the GBs are cached on the ideal, and different for
different orderings. no-one noticed.
Question
  are in your examples the bases supposed tobe identical?

**Comment by jankoboehm** [comment]
One might compute two Gröbner bases through two different methods wrt the same ordering (or per se know that one is a Gröbner basis), and want to compare. So == makes sense, I would say.

This is indeed a question for IdealGens. Two should be == if all entries on the OSCAR side are equal (ring, generating system (in the specified order of elements), monomial ordering, flag isGB, flag isReduced. The flag keep_ordering is as far as I understand historical nonsense and should go. The Singular side of the BiPolyArray contained in the IdealGens should be always consistent with the Oscar side and might not be there in the first place, so is not tested for equality. If there is only a Singular side, we could test there or copy over with oscar_assure, which could anyway always be called.

**Comment by fieker** [comment]
On Thu, Mar 09, 2023 at 12:48:41AM -0800, Janko Boehm wrote:
> One might compute two Gröbner bases through two different methods wrt the same ordering (or per se know that one is a Gröbner basis), and want to compare. So == makes sense, I would say.
> 
> This is indeed a question for IdealGens. Two should be == if all entries on the OSCAR side are equal (ring, generating system (in the specified order of elements), monomial ordering, flag isGB, flag isReduced. The flag keep_ordering is as far as I understand historical nonsense and should go. The Singular side should be always consistent with the Oscar side and might not be there in the first place, so is not tested for equality. If there is only a Singular side, we could test there or copy over with oscar_assure, which could anyway always be called.
> 
Why not just

 length the same
 and the polys the same - regardless of the ordering?

The keep_ordering is tricky: on oscar, a polynomial can only be in one
of the 3 AA orderings. We can pretend to be in others (with the ordering
flag added), but at the end of the day, the polynomials are in Oscar in
one of the 3.
So when converting from Oscar to Singular, for some problems we
definitely want to keep the ordering, for others, like ideal membership,
... we don't

**Comment by jankoboehm** [comment]
To have a more transparent situation, could one not just specify the ordering corresponding to the AA ordering you want to keep and should be known, instead of having the keep_ordering flag? 

On the question on permutation of the generators: One of the key uses of IdealGens is to write elements with respect to the generating system. So if one uses the same coeffcients with a pemutation of the generators, one will typically get the wrong element.

**Comment by fingolfin** [comment]
My suggestion would be to define equality as "same vector of polynomials". That's easy and simple. And it is someone wants to test up to ordering and duplicates, one can just feed both GBs into `Set` instances and compare the result

**Comment by fingolfin** [comment]
Sorry, when I wrote "same polynomials in the same ordering" last night what I meant was "the same *order*", i.e. this is not about the polynomial ordering, but rather the order of the polynomials in the vector. So I don't want to consider the monomial ordering at all.

Sorry for the confusion

**Comment by afkafkafk13** [comment]
Yes, testing for the same vector of polynomials to test equality of ordered lists of generators is precisely the right approach from my point of view

The additional comparisons (monomial ordering, is_GB,...) suggested bei @jankoboehm only make sense in the context of Groebner  bases, where the comparison will mostly be used. However, if the given lists of generators are the same and the monomial ordering is the same, it does not make sense to spot a difference in is_GB (which might by chance not have been set for one of the two, e.g. it the input was already a GB and the naive user just tested this property by computation of a GB of it). Either the list is somehow certified to be a GB (whence we can use this information in the future for both identical copies of the list) or it is not certified to have this property. So we can only trust a 'true' in is_GB.

**Comment by jankoboehm** [comment]
@afkafkafk13, your argument goes in the right direction, but I think not totally. One should probably see it as follows: IdealGens disguises as three different types distinguished through attributes (with this route taken since historically there was only the heavily utilized BiPolyArray as a non-user facing type, which was happily used as any of the three types): an ideal generating system, an ideal generating system together with a monomial ordering, and a Groebner basis. The Groebner basis case is already distinguished user facing though different printing, projecting as a different type. One could also distinguish the second case in the printing. The second case could be disallowed, but is indeed kind of essential, e.g. it can be used to avoid unnecessary sorting by making use of the Singular side, accomodating the lack of full-fledged monomial orderings except the three AA in the polynomial data structures on the Oscar side. From this point of view, how to test equality is then obvious since objects of different data types should not be equal: 

In case one (no monomial ordering set, isGB not set) we check equality of vectors of generators, in the second case (monomial ordering set), we also check the ordering, same in case three (isGB set which implies monomial ordering set), and of course different cases unequal. The attribute isReduced is just additional information a Groebner basis can learn about itself and should not be tested. Note that isGB does not make sense as an attribute to case one, since a Groebner basis requires intrinsically an ordering, so it could only be an attribute to case two, which would lead to certified or not nonsense, which I do not think we want.

**Comment by afkafkafk13** [comment]
Thank you for clarifying this.

Indeed, my argument was on the theoretical side not on the technical one. The distinction between case 1 and the other cases is obvious in theory and practice based on the existence of a prescribed monomial ordering. 

The other two cases, however, only differ technically and I was not aware that this distinction is actively used. For me both situations were ideal generators in the presence of an ordering, which know whether they are certified to be a Groebner basis -- in some way like the old attribute isSB in Singular (which did not influence the printing, but was very handy to be set 'manually' in certain contexts). In other words, I was still too much in the Singular world with my thoughts.

**Comment by fingolfin** [comment]
@jankoboehm will figure out a way forward on this...

**Comment by joschmitt** [comment]
Closed via #2622.
